pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_controller.go

```diff
@@ -0,0 +1,265 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiexportendpointslice
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	corev1alpha1 "github.com/kcp-dev/kcp/pkg/apis/core/v1alpha1"
+	kcpclientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/cluster"
+	apisv1alpha1client "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/typed/apis/v1alpha1"
+	apisinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apis/v1alpha1"
+	corev1alpha1informers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/core/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/indexers"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	"github.com/kcp-dev/kcp/pkg/reconciler/committer"
+)
+
+const (
+	ControllerName = "kcp-apiexportendpointslice"
+
+	DefaultIdentitySecretNamespace = "kcp-system"
+)
+
+// NewController returns a new controller for APIExportEndpointSlices.
+// Shards and APIExports are read from the cache server
+func NewController(
+	apiExportEndpointSliceClusterInformer apisinformers.APIExportEndpointSliceClusterInformer,
+	localShardClusterInformer corev1alpha1informers.ShardClusterInformer,
+	globalShardClusterInformer corev1alpha1informers.ShardClusterInformer,
+	localAPIExportClusterInformer apisinformers.APIExportClusterInformer,
+	globalAPIExportClusterInformer apisinformers.APIExportClusterInformer,
+	kcpClusterClient kcpclientset.ClusterInterface,
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName)
+
+	c := &controller{
+		queue: queue,
+		listAPIExportEndpointSlices: func() ([]*apisv1alpha1.APIExportEndpointSlice, error) {
+			return apiExportEndpointSliceClusterInformer.Lister().List(labels.Everything())
+		},
+		listShards: func() ([]*corev1alpha1.Shard, error) {
+			shards, err := globalShardClusterInformer.Lister().List(labels.Everything())
+			if err != nil {
+				var localErr error
+				shards, localErr = localShardClusterInformer.Lister().List(labels.Everything())
```

Comment:
mhm, I think that `localShardClusterInformer` on the `root` shard will only hold some items.

Unless we want to put the list of shards from the cache into a local configmap :)

I was thinking about plugging and using  `c.ShardVirtualWorkspaceURL()`. I was also thinking about using it in the reconciler loop only for new `APIExportEndpointSlices` (as a fallback) not in the `listShards` func.

Reply:
> mhm, I think that localShardClusterInformer on the root shard will only hold some items.

That's what I have been saying. Today:
- only the root shard holds Shards objects
- it is IMHO sub-optimal (fragile) to have a "special" shard
 
> Unless we want to put the list of shards from the cache into a local configmap :)

we already have a local cache of the remote cache... 

> I was thinking about plugging and using c.ShardVirtualWorkspaceURL()

mind that some filtering through Partition is added in a later PR. Filtering based on bound APIs should also come later. This will be easier with a uniform approach.

> I was also thinking about using it in the reconciler loop only for new APIExportEndpointSlices (as a fallback) not in the listShards func.

I try to keep the pattern that requests are done through lister/getter functions of the controller. Amongst others it makes easier to test the reconciliation logic.

Reply:
ok, I see your point, how about:

- waiting for `waitForOptionalSync` before running the controller
- removing `localShardClusterInformer` 
- keeping `localAPIExportClusterInformer`

that would allow for running the controller when we have seen some data from the cache and would allow to reconcile local APIExports even if the connection with the cache has been lost - in that case we would read what we already have in `globalShardClusterInformer`

Reply:
Putting aside @fgiloux's suggestion to avoid having the root shard be "special" ... @p0lyn0mial why would the `localShardClusterInformer` on `root` only hold a subset of Shards?

Reply:
oh, sorry, my mistake, the `root` shard holds the entire set.

Reply:
@fgiloux since we use `waitForOptionalSync` I think we can remove `localShardClusterInformer`

Reply:
why do we have to look into local shards? Always go through global.